wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginMetaRetriever.java

```diff
@@ -87,11 +93,23 @@ public PluginMetaRetriever(@Nullable @Named(PLUGIN_REGISTRY_PROPERTY) String plu
    */
   @Beta
   public Collection<PluginMeta> get(Map<String, String> attributes) throws InfrastructureException {
-    if (pluginRegistry == null || attributes == null || attributes.isEmpty()) {
+    // Have to check for empty value instead of plain null as it's possible to have empty
+    // plugins attribute
+    String pluginsAttribute =
+        attributes.getOrDefault(Constants.WORKSPACE_TOOLING_PLUGINS_ATTRIBUTE, null);
+    String editorAttribute =
+        attributes.getOrDefault(Constants.WORKSPACE_TOOLING_EDITOR_ATTRIBUTE, null);
+
+    // Check if workspace requests ws.next features with no registry configured
+    if (pluginRegistry == null
+        && (!Strings.isNullOrEmpty(pluginsAttribute) || !Strings.isNullOrEmpty(editorAttribute))) {
+      throw new InfrastructureException(CHE_REGISTRY_MISSING_ERROR);
+    }
+
+    // Check if any plugins/editor is in workspace
+    if (Strings.isNullOrEmpty(editorAttribute) && Strings.isNullOrEmpty(pluginsAttribute)) {
```

Comment:
I think we can simplify all the checks a bit. It is not critical, so up to you whether to apply it.
1. Check if it is WS.NEXT flow (attributes are not null, nor empty, editor or plugins attribute is present and not null)
2. If no then return empty list
3. Otherwise
4. If plugin registry is null throw error
5. Otherwise all the code that is left

Reply:
I've streamlined these checks a bit:

1. Check if attributes null -> return empty list
2. Check if both plugins and editors null or empty -> return empty list
3. Check if registry not configured -> throw exception